common/metagraph_syncer.py

```diff
@@ -0,0 +1,145 @@
+import asyncio
+from concurrent.futures import ThreadPoolExecutor
+import dataclasses
+from dataclasses import field
+from datetime import datetime
+import functools
+import bittensor as bt
+from typing import Dict, List, Callable, Optional
+import threading
+import traceback
+
+from common import utils
+
+
+class MetagraphSyncer:
+    @dataclasses.dataclass
+    class _State:
+        metagraph: Optional[bt.metagraph] = None
+        last_synced_time: Optional[datetime] = None
+        listeners: List = field(default_factory=list)
+
+    def __init__(self, subtensor: bt.subtensor, config: Dict[int, int]):
+        """Constructs a new MetagraphSyncer, that periodically refreshes metagraph defined in the config.
+
+        Args:
+            subtensor (bt.subtensor): The subtensor used to fetch the metagraphs.
+            config (Dict[int, int]): A mapping of netuid to the cadence (in seconds) to sync the metagraph.
+        """
+        self.subtensor = subtensor
+        self.config = config
+        self.metagraph_map: Dict[int, MetagraphSyncer._State] = {
+            netuid: MetagraphSyncer._State() for netuid in config.keys()
+        }
+        self.is_running = False
+        self.done_initial_sync = False
+        self.lock = threading.RLock()
+        self.notify_listener_thrad_pool = ThreadPoolExecutor(max_workers=1)
+
+    def do_initial_sync(self):
+        """Performs an initial sync of all metagraphs.
+
+        Unlike regular syncs, this will not notify listeners of the updated metagraph.
+        """
+        bt.logging.debug("Metagraph syncer do_initial_sync called")
+
+        for netuid in self.config.keys():
+            fn = functools.partial(self.subtensor.metagraph, netuid)
+            with self.lock:
+                metagraph = utils.run_in_subprocess(fn, ttl=120)
+                state = self.metagraph_map[netuid]
+                state.metagraph = metagraph
+                state.last_synced_time = datetime.now()
+
+            bt.logging.debug(f"Successfully loaded metagraph for {netuid}")
+
+        self.done_initial_sync = True
+
+    def start(self):
+        bt.logging.debug("Metagraph syncer start called")
+
+        assert self.done_initial_sync, "Must call do_initial_sync before starting"
+
+        self.is_running = True
+        thread = threading.Thread(target=self._run, daemon=True)
+        thread.start()
+
+    async def _sync_metagraph_loop(self, netuid: int, cadence: int):
+        while self.is_running:
+            # On start, wait cadence before the first sync.
+            await asyncio.sleep(cadence)
+
+            try:
+                # Intentionally block the shared thread so that we only
+                # sync 1 metagraph at a time.
+                metagraph = utils.run_in_subprocess(
+                    functools.partial(self.subtensor.metagraph, netuid), ttl=120
+                )
+                state = None
```

Comment:
Don't we immediately get state from the map in the next line? If that throws an exception then setting state to None here wouldn't help right?

Reply:
State is accessed outside the lock on line 85, so I need to define it outside the lock to keep it in scope.